<?xml version="1.0"?>
<launch>

  <!-- Robot Customization -->
  <arg name="ns"                   default=""      doc="namespace" />
  <arg name="verbose"              default="false" doc="Whether more debug output should be printed." />
  <arg name="unpause"              default="false" doc="Whether the simulation should be unpaused on start." />
  <arg name="headless"             default="false" />
  <arg name="render_offscreen"     default="true"  doc="Whether offscreen rendering should be enabled." />
  <arg name="no_x"                 default=""      doc="Set to true to enable running on a server without X, disabling everything related to OpenGL rendering. (deprecated)"/>
  <arg name="no_render"            default="false" doc="Set to true to disabling rendering on- and off screen. (shorthand for render_offscreen:=false headless:=true)"/>
  <arg name="eval_mode"            default="false" doc="Whether to run mujoco_ros in evaluation mode." />
  <arg name="admin_hash"           default="''"    doc="Hash to verify critical operations in evaluation mode." />
  <arg name="debug"                default="false" doc="Whether to run with gdb." />
  <arg name="debug_server"         default="false" doc="Whether to run with gdbserver on port 1234" />
  <arg name="valgrind"             default="false" doc="Whether to run with valgrind" />
  <arg name="valgrind_args"        default=""      doc="arguments for valgrind" />
  <arg name="wait_for_xml"         default="false" doc="Whether mujoco_ros should wait for an xml in the parameter server." />
  <arg name="realtime"             default=""      doc="Fraction of desired realtime (0,1]. -1 to run as fast as possible." />
  <arg name="profile"              default="false" doc="Whether mujoco_ros should be profiled." />
  <arg name="num_sim_steps"        default="-1" />
  <arg name="mujoco_plugin_config" default=""      doc="Optionally provide the path to a yaml with plugin configurations to load." />
  <arg name="mujoco_threads"       default="1"     doc="Number of threads to use in the MuJoCo simulation." />

  <arg name="modelfile"            default="$(find mujoco_ros)/assets/pendulum_world.xml"        doc="MuJoCo xml file to load. Should define robot model and world." />
  <arg name="initial_joint_states" default="$(find mujoco_ros)/config/initial_joint_states.yaml" doc="Provide a filepath containing initial joint states to load." />
  <arg name="console_config_file"  default="$(find mujoco_ros)/config/rosconsole.config"         doc="Path to ROS console config used when verbose logging is active." />

  <arg name="use_sim_time" />

  <param name="/use_sim_time" value="$(arg use_sim_time)"/>

  <env name="ROSCONSOLE_FORMAT" value="[${severity}] [${time}] [${logger}] [${node}] [${function}]: ${message}"/>
  <env if="$(arg verbose)" name="ROSCONSOLE_CONFIG_FILE"
       value="$(arg console_config_file)"/>

  <group if="$(eval arg('mujoco_plugin_config') != '')">
    <rosparam file="$(arg mujoco_plugin_config)" subst_value="true" />
  </group>

  <group if="$(arg debug)">
    <group unless="$(arg debug_server)">
      <node pkg="mujoco_ros" type="mujoco_node" name="mujoco_server" output="screen" launch-prefix="gdb --args" args="--admin-hash $(arg admin_hash)" >
        <param name="ns"                   value="$(arg ns)" />
        <param name="unpause"              value="$(arg unpause)" />
        <param name="headless"             value="$(arg headless)" />
        <param name="render_offscreen"     value="$(arg render_offscreen)" />
        <param name="no_x"                 value="$(arg no_x)" />
        <param name="no_render"            value="$(arg no_render)" />
        <param name="num_steps"            value="$(arg num_sim_steps)" />
        <param name="eval_mode"            value="$(arg eval_mode)" />
        <param name="modelfile"            value="$(arg modelfile)" />
        <param name="wait_for_xml"         value="$(arg wait_for_xml)" />
        <param name="realtime"             value="$(arg realtime)" />
        <param name="num_mj_threads"       value="$(arg mujoco_threads)" />
        <rosparam file="$(arg initial_joint_states)" subst_value="true" />
      </node>
    </group>
    <group if="$(arg debug_server)">
      <node pkg="mujoco_ros" type="mujoco_node" name="mujoco_server" output="screen" launch-prefix="gdbserver localhost:1234" args="--admin-hash '$(arg admin_hash)'" >
        <param name="ns"                   value="$(arg ns)" />
        <param name="unpause"              value="$(arg unpause)" />
        <param name="headless"             value="$(arg headless)" />
        <param name="render_offscreen"     value="$(arg render_offscreen)" />
        <param name="no_x"                 value="$(arg no_x)" />
        <param name="no_render"            value="$(arg no_render)" />
        <param name="num_steps"            value="$(arg num_sim_steps)" />
        <param name="eval_mode"            value="$(arg eval_mode)" />
        <param name="modelfile"            value="$(arg modelfile)" />
        <param name="wait_for_xml"         value="$(arg wait_for_xml)" />
        <param name="realtime"             value="$(arg realtime)" />
        <param name="num_mj_threads"       value="$(arg mujoco_threads)" />
        <rosparam file="$(arg initial_joint_states)" subst_value="true" />
      </node>
    </group>
  </group>

  <group unless="$(arg debug)">
    <group if="$(arg valgrind)">
      <node pkg="mujoco_ros" type="mujoco_node" name="mujoco_server" output="screen" args="--admin-hash $(arg admin_hash)" launch-prefix="valgrind $(arg valgrind_args)" >
        <param name="ns"                   value="$(arg ns)" />
        <param name="unpause"              value="$(arg unpause)" />
        <param name="headless"             value="$(arg headless)" />
        <param name="render_offscreen"     value="$(arg render_offscreen)" />
        <param name="no_x"                 value="$(arg no_x)" />
        <param name="no_render"            value="$(arg no_render)" />
        <param name="num_steps"            value="$(arg num_sim_steps)" />
        <param name="eval_mode"            value="$(arg eval_mode)" />
        <param name="modelfile"            value="$(arg modelfile)" />
        <param name="wait_for_xml"         value="$(arg wait_for_xml)" />
        <param name="realtime"             value="$(arg realtime)" />
        <param name="num_mj_threads"       value="$(arg mujoco_threads)" />
        <rosparam file="$(arg initial_joint_states)" subst_value="true" />
      </node>
    </group>
    <group unless="$(arg valgrind)">
      <group if="$(arg profile)">
        <env name="CPUPROFILE" value="/tmp/profile.out" />
      </group>
      <node pkg="mujoco_ros" type="mujoco_node" name="mujoco_server" output="screen" args="--admin-hash $(arg admin_hash)" >
        <param name="ns"                   value="$(arg ns)" />
        <param name="unpause"              value="$(arg unpause)" />
        <param name="headless"             value="$(arg headless)" />
        <param name="render_offscreen"     value="$(arg render_offscreen)" />
        <param name="no_x"                 value="$(arg no_x)" />
        <param name="no_render"            value="$(arg no_render)" />
        <param name="num_steps"            value="$(arg num_sim_steps)" />
        <param name="eval_mode"            value="$(arg eval_mode)" />
        <param name="modelfile"            value="$(arg modelfile)" />
        <param name="wait_for_xml"         value="$(arg wait_for_xml)" />
        <param name="realtime"             value="$(arg realtime)" />
        <param name="num_mj_threads"       value="$(arg mujoco_threads)" />
        <rosparam file="$(arg initial_joint_states)" subst_value="true" />
      </node>
    </group>
  </group>

</launch>
